src/communication/base_communicator.hpp

```diff
@@ -0,0 +1,149 @@
+#pragma once
```

Comment:
Don't understand what this is referring to. The name of what is "the generic"? You mean that "communicator" itself is too generic and we should name it something like "internode-communicator" or "domain_communicator"?